pkg/waiter/manager.go

```diff
@@ -0,0 +1,255 @@
+	managerShutdownCh chan struct{}
+
+	// maxCollisions the max number of id generation collisions allowed prior to error
+	// when creating a new waiter.
+	maxCollisions int
+}
+
+// NewManager creates a new waiter Manager.
+func NewManager[T any](opts ...Option) *managerImpl[T] {
+	var options options
+
+	options.maxCollisions = defaultMaxCollisions
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	if options.idGenerator == nil {
+		options.idGenerator = &UUIDGenerator{}
+	}
+
+	return &managerImpl[T]{
+		idGenerator:       options.idGenerator,
+		waiters:           map[string]chan *response[T]{},
+		responseCh:        make(chan *response[T]),
+		doneWaiterCh:      make(chan string),
+		managerShutdownCh: make(chan struct{}),
+		maxCollisions:     options.maxCollisions,
+	}
+}
+
+// Start spawns a goroutine that will run forever or until ctx done delivering
+// messages to waiters.
+func (w *managerImpl[T]) Start(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				// ensure are no more sends.
+				close(w.managerShutdownCh)
+
+				// inform the waiters.
+				w.closeWaiters()
+				return
+			case r := <-w.responseCh:
+				waiterCh, err := w.removeWaiter(r.id)
+				if errors.Is(err, errNoExists) {
+					continue
+				}
+
+				// to prevent blocking, waiterCh should be buffered with
+				// size of 1 (each waiterCh should only ever receive 1 msg)
+				// (alternatively use a goroutine).
+				waiterCh <- r
+			case id := <-w.doneWaiterCh:
+				// the waiter has been closed or canceled.
+				_, _ = w.removeWaiter(id)
+			}
+		}
+	}()
+}
+
+// Send sends data and err to the waiter with the provided id.
+func (w *managerImpl[T]) Send(id string, data T, err error) error {
+	// if the manager is shutdown, return immediately.
+	select {
+	case <-w.managerShutdownCh:
+		return ErrManagerShutdown
+	default:
+	}
+
+	// check again if the manager is shutdown, return if so
+	// otherwise write the message to the proper channel.
+	select {
+	case <-w.managerShutdownCh:
+		return ErrManagerShutdown
+	case w.responseCh <- &response[T]{
+		id:   id,
+		data: data,
+		err:  err,
+	}:
+	}
+
+	return nil
+}
+
+// NewWaiter creates a waiter with a unique ID. A call to Wait() will complete when
+// a response is published using this ID.
+func (w *managerImpl[T]) NewWaiter() (Waiter[T], error) {
+	// if the manager is shutdown, error out.
+	select {
+	case <-w.managerShutdownCh:
+		return nil, ErrManagerShutdown
+	default:
+	}
+
+	// otherwise setup the new waiter.
+	var waiterCh chan *response[T]
+	for i := 0; i < w.maxCollisions; i++ {
+		id, err := w.idGenerator.GenID()
+		if err != nil {
+			return nil, err
+		}
+
+		waiterCh, err = w.addWaiter(id)
+		if err == nil {
+			return newGenericWaiter(id, waiterCh, w.doneWaiterCh), nil
+		}
+	}
+
+	return nil, ErrTooManyCollisions
+}
+
+func (w *managerImpl[T]) addWaiter(id string) (chan *response[T], error) {
+	w.waitersMu.Lock()
+	defer w.waitersMu.Unlock()
```

Comment:
super nit: could first hold a read lock, then only acquire the write lock if the key doesn't exist yet.

Reply:
Ohh Interesting, never considered using both in the same function. Will consider